src/common/lua/ngx_wasm_lua_ffi.c

```diff
@@ -257,3 +258,218 @@ ngx_http_wasm_ffi_set_host_properties_handlers(ngx_http_request_t *r,
     return ngx_proxy_wasm_properties_set_ffi_handlers(pwctx, getter, setter, r);
 }
 #endif
+
+
+void
+ngx_wa_ffi_shm_lock(ngx_wa_shm_t *shm)
+{
+    ngx_wa_assert(shm);
+
+    ngx_wa_shm_lock(shm);
```

Comment:
Not a blocker for this PR, I think we can merge it as such, but reading this makes me realize that all locking APIs in OpenResty also come with a timeout handler that automatically releases the lock on expiration. Given our FFI API is not directly exposed to end-users this is probably fine for now granted we review & ensure correct usage, but a timeout handler and API option could become necessary in the future.

Reply:
Hm, that's interesting. Could you point me to one of these APIs?

I don't see such a timeout handler being employed to release locks in the `shdict` library. See this [excerpt](https://github.com/openresty/lua-nginx-module/blob/master/src/ngx_http_lua_shdict.c#L524-L578) from the `ngx_http_lua_shdict_get_keys`, for example.

Reply:
Only when the lock is being explicitly manipulated by the end-user, not when it is part of the OpenResty codebase itself. E.g. the [semaphore API](https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/semaphore.md#wait), or other lua-resty libraries built on top of locking mechanisms (and exposing these locks to the end-user). Hence as we are not advertising our FFI lib to end-users (only to Gateway developers), I think we can skip this safeguard, but we should still keep it in mind wen reviewing Gateway code using this FFI.

Reply:
Oh, got it, thanks!